RoT/Protocol/RFCs/0000-Capabilities2.md

```diff
@@ -0,0 +1,140 @@
+proposes removing the payload negotiation into the transport layer.
+
+This proposal replaces "Device Capabilities" with an extensible negotiation
+flow that allow arbitrary new capabilities to be exchanged, while respecting the
+maximum 64-byte size of required messages.
+
+# Proposal
+
+This proposal adds one new message, called "Negotiation", with the following
+format:
+
+```
+struct {
+  uint8_t count;
+  uint8_t reserved;
+  // Records...
+}
+```
+
+The `reserved` byte must always be set to zero, and servers must reject
+non-zero values. This may eventually be used to implement continuation bits for
+negotiation, if either side of the connection cannot communicate every
+capability in a single message.
+
+After the reserved byte, an arbitrarily long sequence of `count` *records* may
+follow. A record consists of:
+1. A one-byte tag identifying the capability it communicates, in the range `0x00`
+   to `0x7f`. Tags with the top bit set are reserved for future specification.
+   Either end of the connection must reject tags with the top bit set, since its
+   length may be unspecified.
+2. At least one byte of contents, encoded in one of three forms:
+   - As a six-bit integer, encoded as a byte with the top two bits forming the
+     value `0b00`.
+   - As a fourteen-bit integer, encoded as two bytes, where the value is formed
+     as `contents[0] | (contents[1] << 6)`. The top two bits form the value
+     `0b01`.
+   - As an array of bytes, where the length is encoded in the bottom six bits of
+     the first byte, with the top two bits having the value `0b10`.
+   - As an array of bytes, where the length is encoded in the first two bytes as
+     a fourteen-bit integer, following the scheme above. The top two bits form
+     the value `0b11`.
+
+In practice, we expect only the first two forms to be used, but since we want
+implementations to skip over records they don't recognize, we specify up to
+16-kilobyte records for forward-compatibility.
+
+Records must appear in lexicographic order: implementations must reject
+out-of-order or duplicate records. This simplifies the process for detecting
+duplicate records.
+
+When beginning communications with a remote RoT, an RoT must send a Negotiation
+message with all of the capabilities it wishes to inform the remote device of.
+The remote device may then either respond with its own capabilities, or return
+an error, if the client failed to provide a critical capability required for
+successful communication. The client may then, in turn, discontinue the
+communication if it does not find the server's capabilities satisfactory.
+
+The following capabilities are pre-defined. We do not currently specify a range
+for vendor use.
+
+- "Mode", `0x00`. Six bit record. Required.
+  - Bits [1:0] are the device type. They are interpreted as in the Device
+    Capabilities message (AC-RoT, PA-RoT, etc.).
+  - Bits [3:2] are the bus role (i.e., can it act as a "client" or "server" on
```

Comment:
It seems you are generally trying to move to client/server nomenclature.  Maybe we should discuss and align on appropriate terminology.  My only concern here is the overloading of the term 'server', since the main use-case for Cerberus is in server platforms.

Reply:
Let's stick to host/device; I have a habit of switching to client/server when chatting with folks who work on TLS, but that shouldn't leak into the spec.

I will note that I've intentionally avoided the master/slave nomenclature here.